### ISSUE TITLE
How is the Nearest Neighbor method implemented?

### ISSUE THREAD

**Issue by ccdou** [Yuheng-Li/UniversalFakeDetect#12]
Thank you for your work. 

The paper mentions the use of two methods, Nearest Neighbor and Linear Classification. 
In the published code, I only found the implementation of Linear Classification and couldn't find how Nearest Neighbor is implemented. Is this part not disclosed?

**Comment by 2kxx** [comment]
I found this, too

**Comment by 2kxx** [comment]
Hi, ccdou. I find this problem, too. Do you get the answer?

**Comment by ccdou** [comment]
> Hi, ccdou. I find this problem, too. Do you get the answer?

no,I don't use this method in my work

**Comment by 2kxx** [comment]
> > Hi, ccdou. I find this problem, too. Do you get the answer?
> 
> no,I don't use this method in my work

Okay, it seems I have no choice but to do it myself.

**Comment by yovela2001** [comment]
> Okay, it seems I have no choice but to do it myself.

Hi, have you implemented this feature? If so, would you mind sharing your code? Thank you!

**Comment by 2kxx** [comment]
> > Okay, it seems I have no choice but to do it myself.
> 
> Hi, have you implemented this feature? If so, would you mind sharing your code? Thank you!

May be this one. I wrote the code long long ago, so you have to understand it yourself.
`def validate4(model, loader1, num_list1):

    with torch.no_grad():
        y = []

        print ("Length of dataset: %d" %(len(loader1)))
        for img, label in loader1:
            in_tens = img.cuda()
            for i in model(in_tens, return_feature=True):
                y.append(i)

            # y = [tensor.to(dtype=torch.float32) for tensor in y]

        cosine_distances = torch.zeros(len(y)).cuda()
        cosine_distances += 1

        folder_path = "./buffer/"
        file_list = [f for f in os.listdir(folder_path)]
        for file_path in file_list:
            tensor = torch.load(os.path.join(folder_path, file_path))
            for i in tensor:
                for k, j in enumerate(y):
                    distance = cosine_distance(j, i)
                    distance[torch.isclose(distance, torch.zeros_like(distance), atol=2e-7, rtol=0)] = 1
                    cosine_distances[k] = min(cosine_distances[k], min(distance))
        print(cosine_distances)
        sorted_indices = torch.argsort(cosine_distances)
        sorted_y1 = [num_list1[i] for i in sorted_indices]
        print(sorted_y1)
        with open('t.txt', "w") as file:
            answer = "".join(str(item + ' ') for item in sorted_y1)
            file.write(answer)`
